File: examples/standalone/ts/07-auth-validation.ts
// Example 7: Auth and token validation
// Tests shatter's ability to reason about string structure, JSON parsing,
// and multi-step validation pipelines. These patterns are ubiquitous in
// web backends and security-critical code.

// EXPECTED BRANCHES (10):
//   1. token is empty                            → "invalid: empty token"
//   2. token doesn't have 3 dot-separated parts  → "invalid: malformed token"
//   3. header JSON parse fails                   → "invalid: corrupt header"
//   4. header.alg !== "HS256"                    → "invalid: unsupported algorithm"
//   5. payload JSON parse fails                  → "invalid: corrupt payload"
//   6. payload.exp < now (expired)               → "invalid: token expired"
//   7. payload.iss !== expectedIssuer            → "invalid: wrong issuer"
//   8. payload.aud doesn't include audience      → "invalid: wrong audience"
//   9. payload.scope missing required scope      → "invalid: insufficient scope"
//  10. all checks pass                           → "valid"
//
// DIFFICULTY: Hard. The token must be a well-formed string with exactly
// two dots, where each segment decodes from base64 to valid JSON with
// specific field values. Random guessing essentially never produces this.

export function validateJwt(
    token: string,
    expectedIssuer: string,
    expectedAudience: string,
    requiredScope: string,
    nowEpochSeconds: number
): string {
    if (token.length === 0) {
        return "invalid: empty token";
    }

    const parts = token.split(".");
    if (parts.length !== 3) {
        return "invalid: malformed token";
    }

    let header: { alg?: string };
    try {
        header = JSON.parse(Buffer.from(parts[0], "base64").toString());
    } catch {
        return "invalid: corrupt header";
    }

    if (header.alg !== "HS256") {
        return "invalid: unsupported algorithm";
    }

    let payload: {
        exp?: number;
        iss?: string;
        aud?: string | string[];
        scope?: string;
    };
    try {
        payload = JSON.parse(Buffer.from(parts[1], "base64").toString());
    } catch {
        return "invalid: corrupt payload";
    }

    if (typeof payload.exp === "number" && payload.exp < nowEpochSeconds) {
        return "invalid: token expired";
    }

    if (payload.iss !== expectedIssuer) {
        return "invalid: wrong issuer";
    }

    const audiences = Array.isArray(payload.aud)
        ? payload.aud
        : [payload.aud];
    if (!audiences.includes(expectedAudience)) {
        return "invalid: wrong audience";
    }

    const scopes = (payload.scope ?? "").split(" ");
    if (!scopes.includes(requiredScope)) {
        return "invalid: insufficient scope";
    }

    return "valid";
}

// Role-based access control.
//
// EXPECTED BRANCHES (12):
//   1. user.roles is empty                       → "denied: no roles"
//   2. user.active === false                     → "denied: inactive user"
//   3. resource is empty string                  → "denied: invalid resource"
//   4. action not in allowed actions             → "denied: invalid action"
//   5. user has "superadmin" role                → "granted: superadmin"
//   6. user is resource owner AND action is "read" → "granted: owner-read"
//   7. user is resource owner AND action is "write" → "granted: owner-write"
//   8. user is resource owner AND action is "delete" → "granted: owner-delete"
//   9. user has "admin" role AND action is "read" → "granted: admin-read"
//  10. user has "admin" role AND action is "write" → "granted: admin-write"
//  11. user has "admin" role AND action is "delete" → "denied: admin-no-delete"
//  12. user has only "viewer" role               → action === "read" ? "granted: viewer" : "denied: viewer-readonly"
//
// DIFFICULTY: Medium-hard. Requires specific combinations of roles, ownership
// status, and actions. The priority ordering (superadmin > owner > admin > viewer)
// means the solver must understand which checks happen first.

interface User {
    id: string;
    roles: string[];
    active: boolean;
}

const VALID_ACTIONS = ["read", "write", "delete"] as const;
type Action = typeof VALID_ACTIONS[number];

export function authorizeRequest(
    user: User,
    resource: string,
    resourceOwnerId: string,
    action: string
): string {
    if (user.roles.length === 0) {
        return "denied: no roles";
    }

    if (!user.active) {
        return "denied: inactive user";
    }

    if (resource.length === 0) {
        return "denied: invalid resource";
    }

    if (!VALID_ACTIONS.includes(action as Action)) {
        return "denied: invalid action";
    }

    if (user.roles.includes("superadmin")) {
        return "granted: superadmin";
    }

    const isOwner = user.id === resourceOwnerId;

    if (isOwner) {
        if (action === "read") {
            return "granted: owner-read";
        }
        if (action === "write") {
            return "granted: owner-write";
        }
        if (action === "delete") {
            return "granted: owner-delete";
        }
    }

    if (user.roles.includes("admin")) {
        if (action === "read") {
            return "granted: admin-read";
        }
        if (action === "write") {
            return "granted: admin-write";
        }
        return "denied: admin-no-delete";
    }

    if (user.roles.includes("viewer")) {
        if (action === "read") {
            return "granted: viewer";
        }
        return "denied: viewer-readonly";
    }

    return "denied: no matching role";
}

File: examples/standalone/rust/07_auth_validation.rs
// Example 7: Auth and token validation.
// Tests reasoning about string structure, base64 decoding, and multi-step validation.

use std::collections::HashMap;

/// validate_jwt — 10 branches: empty token→"invalid: empty token",
/// not 3 parts→"invalid: malformed token", bad header base64→"invalid: corrupt header",
/// alg!="HS256"→"invalid: unsupported algorithm", bad payload base64→"invalid: corrupt payload",
/// expired→"invalid: token expired", wrong issuer→"invalid: wrong issuer",
/// wrong audience→"invalid: wrong audience", missing scope→"invalid: insufficient scope",
/// all pass→"valid".
///
/// Simplified: uses naive key=value pairs instead of real JSON/base64 to keep stdlib-only.
fn validate_jwt(
    token: &str,
    expected_issuer: &str,
    expected_audience: &str,
    required_scope: &str,
    now_epoch_seconds: i64,
) -> &'static str {
    if token.is_empty() {
        return "invalid: empty token";
    }

    let parts: Vec<&str> = token.split('.').collect();
    if parts.len() != 3 {
        return "invalid: malformed token";
    }

    let header = match parse_simple_kv(parts[0]) {
        Some(m) => m,
        None => return "invalid: corrupt header",
    };

    if header.get("alg").map(|s| s.as_str()) != Some("HS256") {
        return "invalid: unsupported algorithm";
    }

    let payload = match parse_simple_kv(parts[1]) {
        Some(m) => m,
        None => return "invalid: corrupt payload",
    };

    if let Some(exp_str) = payload.get("exp") {
        if let Ok(exp) = exp_str.parse::<i64>() {
            if exp < now_epoch_seconds {
                return "invalid: token expired";
            }
        }
    }

    if payload.get("iss").map(|s| s.as_str()) != Some(expected_issuer) {
        return "invalid: wrong issuer";
    }

    let aud_match = payload.get("aud").map_or(false, |aud| {
        aud.split(',').any(|a| a.trim() == expected_audience)
    });
    if !aud_match {
        return "invalid: wrong audience";
    }

    let scope_match = payload.get("scope").map_or(false, |scope| {
        scope.split(' ').any(|s| s == required_scope)
    });
    if !scope_match {
        return "invalid: insufficient scope";
    }

    "valid"
}

/// Parses "key1=val1;key2=val2" into a HashMap. Returns None on malformed input.
fn parse_simple_kv(s: &str) -> Option<HashMap<String, String>> {
    if s.is_empty() {
        return None;
    }
    let mut map = HashMap::new();
    for pair in s.split(';') {
        let parts: Vec<&str> = pair.splitn(2, '=').collect();
        if parts.len() != 2 {
            return None;
        }
        map.insert(parts[0].trim().to_string(), parts[1].trim().to_string());
    }
    Some(map)
}

struct User {
    id: String,
    roles: Vec<String>,
    active: bool,
}

const VALID_ACTIONS: &[&str] = &["read", "write", "delete"];

/// authorize_request — 12 branches: no roles→denied, inactive→denied,
/// empty resource→denied, invalid action→denied, superadmin→granted,
/// owner+read→granted, owner+write→granted, owner+delete→granted,
/// admin+read→granted, admin+write→granted, admin+delete→denied,
/// viewer+read→granted, viewer+other→denied, no matching role→denied.
fn authorize_request(
    user: &User,
    resource: &str,
    resource_owner_id: &str,
    action: &str,
) -> &'static str {
    if user.roles.is_empty() {
        return "denied: no roles";
    }
    if !user.active {
        return "denied: inactive user";
    }
    if resource.is_empty() {
        return "denied: invalid resource";
    }
    if !VALID_ACTIONS.contains(&action) {
        return "denied: invalid action";
    }

    if user.roles.iter().any(|r| r == "superadmin") {
        return "granted: superadmin";
    }

    let is_owner = user.id == resource_owner_id;

    if is_owner {
        return match action {
            "read" => "granted: owner-read",
            "write" => "granted: owner-write",
            "delete" => "granted: owner-delete",
            _ => "denied: invalid action",
        };
    }

    if user.roles.iter().any(|r| r == "admin") {
        return match action {
            "read" => "granted: admin-read",
            "write" => "granted: admin-write",
            _ => "denied: admin-no-delete",
        };
    }

    if user.roles.iter().any(|r| r == "viewer") {
        if action == "read" {
            return "granted: viewer";
        }
        return "denied: viewer-readonly";
    }

    "denied: no matching role"
}

fn main() {
    let token = "alg=HS256.iss=myapp;aud=web;scope=admin read;exp=9999999999.signature";
    println!("{}", validate_jwt(token, "myapp", "web", "read", 1000));

    let user = User {
        id: "u1".to_string(),
        roles: vec!["admin".to_string()],
        active: true,
    };
    println!("{}", authorize_request(&user, "/docs", "u2", "read"));
}
